File: app/src/test/java/com/coinninja/coinkeeper/util/android/PreferencesUtilTest.java
package com.coinninja.coinkeeper.util.android;

import android.content.SharedPreferences;

import com.coinninja.coinkeeper.TestCoinKeeperApplication;
import com.coinninja.coinkeeper.interactor.PreferenceInteractor;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static junit.framework.TestCase.assertTrue;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


@RunWith(RobolectricTestRunner.class)
@Config(application = TestCoinKeeperApplication.class)
public class PreferencesUtilTest {

    @Mock
    SharedPreferences sharedPreferences;

    @Mock
    SharedPreferences.Editor editor;

    @Mock
    PreferencesUtil.Callback callback;

    private PreferencesUtil preferencesUtil;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        when(sharedPreferences.edit()).thenReturn(editor);

        preferencesUtil = new PreferencesUtil(sharedPreferences);
    }

    @Test
    public void allows_write_of_string_preference() {
        String value = "value";
        String key = "key";

        preferencesUtil.savePreference(key, value);

        verify(editor).putString(key, value);
        verify(editor).apply();
    }

    @Test
    public void allows_write_of_boolean_preference_with_callback() {
        String key = "key";
        boolean value = false;
        PreferencesUtil.PreferenceWriter preferenceWriter = mock(PreferencesUtil.PreferenceWriter.class);
        preferencesUtil.preferenceWriter = preferenceWriter;
        when(preferenceWriter.clone()).thenReturn(preferenceWriter);
        when(preferenceWriter.setPreference(editor, callback, key, value)).thenReturn(preferenceWriter);

        preferencesUtil.savePreference(key, value, callback);

        verify(preferenceWriter).clone();
        verify(preferenceWriter).setPreference(editor, callback, key, value);
        verify(preferenceWriter).execute();
    }

    @Test
    public void preference_util_saves_long_value_to_shared_pref_test() {
        long sampleLongValue = 5546546l;
        String sampleKey = "SOME - KEY";

        preferencesUtil.savePreference(sampleKey, sampleLongValue);

        verify(editor).putLong(sampleKey, sampleLongValue);
        verify(editor).apply();
    }

    @Test
    public void preference_util_retrieves_long_value_to_shared_pref_test() {
        long sampleLongValue = 5546546l;
        long sampleDefaultValue = 0l;
        String sampleKey = "SOME - KEY";
        when(sharedPreferences.getLong(sampleKey, sampleDefaultValue)).thenReturn(sampleLongValue);

        long savedValue = preferencesUtil.getLong(sampleKey, sampleDefaultValue);

        assertThat(savedValue, equalTo(sampleLongValue));
    }

    @Test
    public void can_save_boolean_to_shared_prefereneces() {
        preferencesUtil.savePreference(PreferenceInteractor.PREFERENCE_FINGERPRINT, true);

        verify(editor).putBoolean(PreferenceInteractor.PREFERENCE_FINGERPRINT, true);
        verify(editor).apply();
    }

    @Test
    public void removes_shared_preference_by_key_value() {
        String key = "---- some key";

        preferencesUtil.removePreference(key);

        verify(editor).remove(key);
        verify(editor).apply();
    }

    @Test
    public void can_retrieve_boolean_from_preferences() {
        when(sharedPreferences.getBoolean(PreferenceInteractor.PREFERENCE_FINGERPRINT, false)).thenReturn(true);

        assertTrue(preferencesUtil.getBoolean(PreferenceInteractor.PREFERENCE_FINGERPRINT));

    }

    @Test
    public void commits_preference_asyncronoslly() {
        PreferencesUtil.PreferenceWriter preferenceWriter = new PreferencesUtil.PreferenceWriter();
        preferenceWriter.editor = editor;
        preferenceWriter.callback = callback;

        preferenceWriter.doInBackground();

        verify(editor).apply();
    }

    @Test
    public void notifies_caller_that_write_complete() {
        PreferencesUtil.PreferenceWriter preferenceWriter = new PreferencesUtil.PreferenceWriter();
        preferenceWriter.editor = editor;
        preferenceWriter.callback = callback;

        preferenceWriter.onPostExecute(null);

        verify(callback).onComplete();
    }

    @Test
    public void allows_task_to_be_cloned() {
        PreferencesUtil.PreferenceWriter prefererenceWriter = new PreferencesUtil.PreferenceWriter();

        assertTrue(prefererenceWriter.clone() instanceof PreferencesUtil.PreferenceWriter);
    }


    @Test
    public void preference_writer_can_write_booleans() {
        String key = "key";
        boolean value = false;

        PreferencesUtil.PreferenceWriter preferenceWriter = new PreferencesUtil.PreferenceWriter();
        preferenceWriter.setPreference(editor, callback, key, value);

        verify(editor).putBoolean(key, value);
        assertThat(preferenceWriter.editor, equalTo(editor));
        assertThat(preferenceWriter.callback, equalTo(callback));

    }

    @Test
    public void fetch_string_from_local_storage() {
        when(sharedPreferences.getString("FOO", "")).thenReturn("bar");

        assertThat(preferencesUtil.getString("FOO", ""), equalTo("bar"));
    }

    @Test
    public void wraps_knowlege_of_presence_of_key() {
        String key = "foo";
        preferencesUtil.contains(key);
        verify(sharedPreferences).contains(key);
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/util/android/PreferencesUtilTest.kt
package com.coinninja.coinkeeper.util.android

import android.content.SharedPreferences
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.coinkeeper.interactor.UserPreferences
import com.nhaarman.mockitokotlin2.whenever
import junit.framework.TestCase.assertTrue
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify


@RunWith(AndroidJUnit4::class)
class PreferencesUtilTest {

    fun setUp(): PreferencesUtil {
        val preferences = mock(SharedPreferences::class.java)
        whenever(preferences.edit()).thenReturn(mock(SharedPreferences.Editor::class.java))
        return PreferencesUtil(preferences)
    }

    @Test
    fun `allows write of string preference`() {
        val util = setUp()
        val value = "value"
        val key = "key"

        util.savePreference(key, value)

        verify(util.editor).putString(key, value)
        verify(util.editor).apply()
    }

    @Test
    fun `allows write of boolean preference with callback`() {
        val util = setUp()
        val key = "key"
        val value = false
        val preferenceWriter = mock(PreferencesUtil.PreferenceWriter::class.java)
        val callback = mock(PreferencesUtil.Callback::class.java)
        util.preferenceWriter = preferenceWriter
        whenever(preferenceWriter.clone()).thenReturn(preferenceWriter)
        whenever(preferenceWriter.setPreference(util.editor, callback, key, value)).thenReturn(preferenceWriter)

        util.savePreference(key, value, callback)

        verify(preferenceWriter).clone()
        verify(preferenceWriter).setPreference(util.editor, callback, key, value)
        verify(preferenceWriter).execute()
    }

    @Test
    fun `preference util saves long value to shared pref test`() {
        val util = setUp()
        val sampleLongValue = 5546546L
        val sampleKey = "SOME - KEY"

        util.savePreference(sampleKey, sampleLongValue)

        verify(util.editor).putLong(sampleKey, sampleLongValue)
        verify(util.editor).apply()
    }

    @Test
    fun `preference util retrieves long value to shared pref test`() {
        val util = setUp()
        val sampleLongValue = 5546546L
        val sampleDefaultValue = 0L
        val sampleKey = "SOME - KEY"
        whenever(util.preferences.getLong(sampleKey, sampleDefaultValue)).thenReturn(sampleLongValue)

        val savedValue = util.getLong(sampleKey, sampleDefaultValue)

        assertThat(savedValue, equalTo(sampleLongValue))
    }

    @Test
    fun `can save boolean to shared prefereneces`() {
        val util = setUp()
        util.savePreference(UserPreferences.PREFERENCE_FINGERPRINT, true)

        verify(util.editor).putBoolean(UserPreferences.PREFERENCE_FINGERPRINT, true)
        verify(util.editor).apply()
    }

    @Test
    fun `removes shared preference by key value`() {
        val util = setUp()
        val key = "---- some key"

        util.removePreference(key)

        verify(util.editor).remove(key)
        verify(util.editor).apply()
    }

    @Test
    fun `can retrieve boolean from preferences`() {
        val util = setUp()
        whenever(util.preferences.getBoolean(UserPreferences.PREFERENCE_FINGERPRINT, false)).thenReturn(true)

        assertTrue(util.getBoolean(UserPreferences.PREFERENCE_FINGERPRINT))

    }

    @Test
    fun `commits preference asyncronoslly`() {
        val util = setUp()
        val preferenceWriter = PreferencesUtil.PreferenceWriter()
        val callback = mock(PreferencesUtil.Callback::class.java)
        preferenceWriter.editor = util.editor
        preferenceWriter.callback = callback

        preferenceWriter.doInBackground()

        verify(util.editor).apply()
    }

    @Test
    fun `notifies caller that write complete`() {
        val util = setUp()
        val preferenceWriter = PreferencesUtil.PreferenceWriter()
        val callback = mock(PreferencesUtil.Callback::class.java)
        preferenceWriter.editor = util.editor
        preferenceWriter.callback = callback

        preferenceWriter.onPostExecute(null)

        verify<PreferencesUtil.Callback>(callback).onComplete()
    }

    @Test
    fun `allows task to be cloned`() {
        val preferenceWriter = PreferencesUtil.PreferenceWriter()

        assertTrue(preferenceWriter.clone() is PreferencesUtil.PreferenceWriter)
    }

    @Test
    fun `preference writer can write booleans`() {
        val util = setUp()
        val key = "key"
        val value = false
        val callback = mock(PreferencesUtil.Callback::class.java)

        val preferenceWriter = PreferencesUtil.PreferenceWriter()
        preferenceWriter.setPreference(util.editor, callback, key, value)

        verify(util.editor).putBoolean(key, value)
        assertThat(preferenceWriter.editor, equalTo(util.editor))
        assertThat(preferenceWriter.callback, equalTo(callback))

    }

    @Test
    fun `fetch string from local storage`() {
        val util = setUp()
        whenever(util.preferences.getString("FOO", "")).thenReturn("bar")

        assertThat(util.getString("FOO", ""), equalTo("bar"))
    }

    @Test
    fun `wraps knowledge of presence of key`() {
        val util = setUp()
        val key = "foo"
        util.contains(key)
        verify(util.preferences).contains(key)
    }

    @Test
    fun `clears preferences`() {
        val util = setUp()

        util.removeAll()

        verify(util.editor).clear()
        verify(util.editor).apply()
    }
}